Add IO encoding function and structs

erickguan opened the pull request:
Fixes #139

matsadler reviewed (commented):
Thank you for this. Sorry for taking so long to get back to you on this.

erickguan commented on src/io.rs:
Thanks — my reading of your comment is that consistency within `magnus` and conciseness are the main goals. It also sounds like you're not opposed to introducing a new `Io` module.

Given that, I suggest we go with `extract_modeenc` — it's concise, and we'd implement it as `Io::extract_modeenc`.

If we take this route, I could also rename `Io::IoEncoding` to `Io::Encoding`. That would better align the module with the C API naming, even if it feels a bit awkward. In practice, only this function (and maybe one or two others) will use or return an encoding anyway.

Reference: [Ruby IO C API](https://docs.ruby-lang.org/capi/en/master/dd/d83/include_2ruby_2io_8h.html#aa22d40106004fd17a113253c6ec79427)

This C API is relatively new, but it carries a lot of legacy from Ruby's history. For example, MRI still uses `enc1` and `enc2` as names for transcoders — a holdover from the pre-UTF days. Pretty wild to see how far some of this stuff stretches back.

matsadler replied:
Thinking it through I think my preference would be renaming this to `io_extract_modeenc`.

This closely matches the C API name, just dropping the redundant `rb_` prefix. The function needs to take `&Ruby` as an argument, so the caller can 'prove' they are on a Ruby thread. It's probably most ergonomic to do this by defining the function as a method on `Ruby` taking `&self`, rather than a free function in the `io` module that takes with `_: &Ruby` as an argument, so `io_` should stay as part of the name.

`IoEncoding` can drop the `Io` prefix as it's in the `io` module.

Sorry for the back and forth, I probably should have given a bit more direction in my first comment rather than just my thoughts on how I've named things in Magnus.

erickguan replied:
Makes sense to me. I updated the function name.

---

Opening the convention book again. If we have `io` prefix, maybe it's still easier to understand with `io::IoEncoding` rather than `io::Encoding`. The current doc looks like this:

![image](https://github.com/user-attachments/assets/df9cc981-10c5-461e-8473-84614e89f2c3)

rustdoc doesn't show a difference in the namespace.

matsadler replied:
I'm happy to leave the `io::IoEncoding`/`io::Encoding` decision up to you.

erickguan commented on src/io.rs:
Updated.

Yeah, it was quite wild to have `enc` and `enc2` as names.

erickguan reviewed (commented):
Thank you so much for reviewing! I understand you might have some other priorities.

A question though about naming before I make more updates to the PR.

matsadler commented:
You might need to rebase your branch on `main` to resolve the CI failures (see #146).

ianks commented:
Can we gate this behind a feature flag? There's are some flags, etc here which I'm unsure about the stability of. With a feature flag, it would allow this code to be used without breaking ruby-head builds in the case something gets changed upstream.

erickguan commented:
Yes, I put in as `io` feature.

erickguan commented:
I adjusted the workspace and crate settings to run tests against this new module. Happy to make further adjustments in this PR or another.

The horror story I found was probably `File.new` interprets the result from `rb_io_extract_modeenc`.  But it's irrelevant in this PR. Ruby internal is surprising say the least.

matsadler commented:
Could you add `io` to the default features. The current features are all features either because they bring in extra dependencies (e.g. `bytes`) or I want to discourage use (e.g. the APIs for interacting with `rb-sys` allow you to break all the promises Magnus tries to make). I don't think the `io` use case falls into either of these.

The `Cargo.toml` changes in 9976817 and 2e941e9 break my workflow for running tests with `cargo test` (without additional arguments) and [Zed](https://zed.dev)'s test runner. Is this change required? Just specifying the `io` feature in the [list of dev dependencies](https://github.com/matsadler/magnus/blob/b83762a95462fa87a1fae445f243e7eb9ae5c8ab/Cargo.toml#L42-L47) seems to work ok when I try it. I'd also prefer not to have a `test-features` feature that leaks into the public API.

erickguan commented:
Thanks for the detailed review. I fixed both points:

- added `io` features in the default feature.
- I reverted the changes to workspaces.

> The Cargo.toml changes in https://github.com/matsadler/magnus/commit/9976817f22f08b35dbace8a3180ffe31bd48ced4 and https://github.com/matsadler/magnus/commit/2e941e96c72430a3eabc041b46c0336c85b64121 break my workflow for running tests with cargo test (without additional arguments) and [Zed](https://zed.dev/)'s test runner. Is this change required? Just specifying the io feature in the [list of dev dependencies](https://github.com/matsadler/magnus/blob/b83762a95462fa87a1fae445f243e7eb9ae5c8ab/Cargo.toml#L42-L47) seems to work ok when I try it. I'd also prefer not to have a test-features feature that leaks into the public API.

You are right my changes are too broad.

I experimented a bit. I can run `cargo test --workspace` for both:

```rust
# `main` branch:
[dev-dependencies]
magnus = { path = ".", default-features = false, features = [
    "embed",
    "rb-sys",
    "bytes",
    "chrono",
] }

# in this PR:
[dev-dependencies]
magnus = { path = ".", default-features = false, features = [
    "embed",
    "rb-sys",
    "bytes",
    "chrono",
    "io",
] }
```

I assume workspace uses the main crate.

Windows build fails for other reasons. I added an intermediate PR #152 to improve the situation.

matsadler commented:
Reran and it's all good now.

Thank you for this contribution, sorry it took so long to get it merged.

erickguan commented:
Thanks for merging.
